Update unit tests draw.ellipse, add unit tests draw.(aa)line(s)

amosbastian opened the pull request:
This pull request contains the following:

### Ellipse

I updated the draw.ellipse unit test so that it now also includes a test that checks if at least two sides of an ellipse, that has a width and height that is 1 pixel smaller than the surface, is touching the surface's border.

It also cleans up the code slightly, including a better way of retrieving the border's values and adds reusability in the form of additional functions.

See: #233

### Lines

It also adds unit tests for draw.aaline, draw.line, draw.aalines and draw.lines.

For draw.aaline and draw.line it now draws a line of differing colours on all types of surfaces and checks if the colour stays the same at position (0, 0). There is also a test that checks if the line contains any gaps, as I noticed aaline(s) sometimes does (or is not drawn at all when close to a surface's border).

The same is done for draw.aalines and draw.lines, except these tests draw a line around the border of the given surface then check if all pixels are of the given colour, or if they contain gaps.

See: #395

illume commented on test/draw_test.py:
```diff
@@ -1,7 +1,325 @@
-from pygame.tests import test_utils
+#################################### IMPORTS ###################################
+
+if __name__ == '__main__':
```
This is_pygame_pkg stuff should be removed.

The imports should be something like:
```python
import unittest

import pygame
from pygame import draw
from pygame.locals import SRCALPHA
from pygame.tests import test_utils
```

e1000 commented on test/draw_test.py:
```diff
@@ -1,7 +1,309 @@
+                color in border for border in borders].count(True)
+            self.assertEqual(sides_touching, 2)
+
+        for width, height in sizes:
```
a loop over `sizes` which contains only one element ? bit strange...

amosbastian replied:
Thanks for pointing that out! I forgot I had removed the other sizes... added them back in.

e1000 commented:
Personnally, I have some issues about *line_is_color* ,*line_has_gaps* and *lines_have_gaps* : 

1. all repeat the  `if line_type == "aalines":` , hence it seems to be repeated code
2. it mixes doing something and testing something , imho, that's bit confusing, it looks bit as you were testing your own test functions ;)
3. ,*line_has_gaps* and *lines_have_gaps* : with a quick look I can not understand what the gaps are ? And why does the equalities at the end mean, there is a gap (or are gaps) ?

So , for me, these helper functions contain a bit too much logic / factorisation ?

e1000 commented on test/draw_test.py:
```diff
@@ -1,7 +1,309 @@
+      # if true, a line will be draw between the first and last points. The
+      # boolean blend argument set to true will blend the shades with
+      # existing shades instead of overwriting them. This function accepts
+      # floating point values for the end points.
```
@illume : should we keep or should we delete  copies of 2008 docstrings ?

illume commented:
@e1000 some good suggestions. Thank you.

1. `line_is_color` and friends could be improved by taking a draw function as the argument instead of a string. Maybe call the argument `draw_line` and then pass in `draw.line` or `draw.aaline`.

2. Then when using line_is_color, the draw_line can be just another parameter in the one test function(test_line_color). This would reduce the number of tests you have too. Also the implementation of `line_is_color` could be moved inside that one test(because now there are not two). So it would be easier for the person reading it to see what it is doing. Because then `line_is_color` and `test_line_color` would fit in one screen.

These two changes would reduce the amount of repeated code (DRY) https://en.wikipedia.org/wiki/Don%27t_repeat_yourself And I think it would bring related test code closer together in the file, making it easier to understand.

3. Yeah, a different explanation of what gaps are would improve this. It would also help readers of the code to put a link to the issue in the docstring. So they could see the image of the issue themselves.

illume commented:
On Travis CI tests I see this error come up:

```
======================================================================
ERROR: test_line (pygame.tests.draw_test.DrawModuleTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/pygame-1.9.5.dev0-py2.7-linux-x86_64.egg/pygame/tests/draw_test.py", line 312, in setUp
    self.surf = pygame.Surface(self.surf_size, pygame.SRCALPHA)
ValueError: no standard masks exist for given bitdepth with alpha
```

So maybe we need to specify the bit depth for when the tests run like this.

amosbastian commented:
Thanks for the feedback @e1000 @illume, I just refactored everything and updated the `lines_set_up()` function so it creates surfaces of an even width/height and odd width/height, with differing depths (16 and 32). If there is anything else left to be improved then please let me know.

e1000 commented on test/draw_test.py:
```diff
@@ -1,7 +1,204 @@
+
+            draw_line(surface, color, (0, 0), (width - 1, 0))
+
+            colors = [surface.get_at((width, 0)) for width in range(width)]
```
Hello, thank you for modifications ! Some more remarks : 

1. it is not so nice to give same name for loop variable `width` and the outer `width` ; better use a single letter for loop variable in list-comprehension : 
```python
colors = [surface.get_at((x, 0)) for x in range(width)]
```
2. In the cases below, you can write `in_border.count(True) == 4` as `all(in_border)` which seems clearer to me.

3. Also, I still don't understand : for me reading the code, *line_has_gaps(* checks that there is *no* gap since _all_ are the same color ... (same remark for below *lines_have_gaps* ) ; could you please explain what the "gaps" are , or am I missing something ?

amosbastian replied:
1. Will update that.
2. Will update that.
3. Since it wasn't clear before I already created #512 showing what I meant.

e1000 commented on test/draw_test.py:
```diff
@@ -1,7 +1,204 @@
+            draw_lines(surface, color, True, points)
+
+            borders = get_border_values(surface, width, height)
+            return [color in border for border in borders]
```
Here also little question : `color in border` tests that the color appears at least one time on the border ;  if I use *any* (he is a good friend of *all*), it is equivalent to : 

```python
return  [any(c == color for c in border) for border in borders]

# but shouldn't it be ?
return  [all(c == color for c in border) for border in borders]
```

amosbastian replied:
I see what you mean, I will also update that.

e1000 commented on test/draw_test.py:
```diff
@@ -1,7 +1,204 @@
+            draw_lines(surface, color, True, points)
+
+            borders = get_border_values(surface, width, height)
+            return [len(border) == border.count(color) for border in borders]
```
I suggest to replace `len(border) == border.count(color)` with the pattern

```python
all(c == color for c in border)
```

amosbastian replied:
Will also update this, thanks.

illume commented:
Wonderful! 🎈 🎉
